### PR TITLE
Add Anti-Raid capabilities

### PR DESCRIPTION
**Pull request by VelvetToroyashi**
Adds anti-raid capabilities to Silk! See [tracker issue](https://issues.silkbot.cc/issue/SBP-22).

### CODE REVIEW

File: src/Silk/Commands/Server/Config/ConfigCommands.Edit.Raid.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            }
+            
+            if (decay is < 10)
+            {
+                await _channels.CreateMessageAsync(_context.ChannelID, "A raid decay of less than ten seconds is not recommended.");
+                return await _channels.CreateReactionAsync(_context.ChannelID, _context.MessageID, $"_:{Emojis.WarningId}");
+            }
+            
+            await _mediator.Send(new UpdateGuildConfig.Request(_context.GuildID.Value)
+                                 {
```
Comment (OoLunar) [inline src/Silk/Commands/Server/Config/ConfigCommands.Edit.Raid.cs]: Weird formatting

File: src/Silk/Commands/Server/Config/ConfigCommands.cs [hunk 2]
```diff
@@ -131,14 +131,19 @@ public async Task<IResult> ViewAllAsync()
...
+               .AppendLine("**Exemptions** | `exemptions`")
                .AppendLine($"> {(config.Exemptions.Any() ? $"Configured AutoMod Exemptions: {config.Exemptions.Count}" : "Not configured")}")
                .AppendLine()
-               .AppendLine($"**Anti-Phishing** | `phishing`")
+               .AppendLine("**Anti-Phishing** | `phishing`")
                .AppendLine($"> {(config.DetectPhishingLinks ? Emojis.EnabledEmoji : Emojis.DisabledEmoji)} {Emojis.WarningEmoji} Detect Phishing Links")
                .AppendLine($"> {(config.BanSuspiciousUsernames ? Emojis.EnabledEmoji : Emojis.DisabledEmoji)} {Emojis.ScanEmoji} Ban Suspicious Usernames")
                .AppendLine($"> {(config.DeletePhishingLinks ? Emojis.EnabledEmoji : Emojis.DisabledEmoji)} {Emojis.DeleteEmoji} Delete Phishing Links")
-               .AppendLine($"> {(action is not null ? Emojis.EnabledEmoji : Emojis.DisabledEmoji)} {Emojis.WrenchEmoji} Post-detection action: {phishingAction}");
+               .AppendLine($"> {(action is not null ? Emojis.EnabledEmoji : Emojis.DisabledEmoji)} {Emojis.WrenchEmoji} Post-detection action: {phishingAction}")
```
Comment (OoLunar) [inline src/Silk/Commands/Server/Config/ConfigCommands.cs:147]: AppendFormat could possibly be used here to increase readability. Unsure if using AppendFormat would increase performance due to .NET 6's new API's regarding interpolated strings: https://devblogs.microsoft.com/dotnet/string-interpolation-in-c-10-and-net-6/

File: src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    
+    private record Raider(Snowflake GuildID, Snowflake RaiderID, string  Reason);
+    
+    private record MessageDTO(Snowflake GuildID, Snowflake ChannelID, Snowflake MessageID, Snowflake AuthorID, int HashCode);
+    
+    private readonly Channel<Raider> _raiders = Channel.CreateUnbounded<Raider>();
+    
+    private readonly ConcurrentDictionary<Snowflake, List<JoinRange>>                      _joinRanges     = new();
+    private readonly ConcurrentDictionary<Snowflake, (DateTimeOffset LastJoin, int Count)> _joins          = new();
+    private readonly ConcurrentDictionary<Snowflake, List<MessageDTO>>               _messageBuckets = new();
```
Comment (OoLunar) [inline src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs]: Inconsistent formatting

File: src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+    public RaidDetectionService(IInfractionService infractions, IDiscordRestUserAPI users, GuildConfigCacheService config)
+    {
+        _infractions = infractions;
+        _users       = users;
+        _config      = config;
+    }
+
+    public async Task<Result> HandleMessageAsync(Snowflake guildID, Snowflake channelID, Snowflake messageID, Snowflake AuthorID, string? content)
```
Comment (OoLunar) [inline src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs]: Inconsistent parameter naming (AuthorId)

File: src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            foreach (var raider in group)
+                await _raiders.Writer.WriteAsync(new Raider(raider.GuildID, raider.AuthorID, "Suspected raid: Coordinated message raid detected."));
+            
+            bucket.RemoveAll(r => group.Contains(r));
+        }
+        
+        return Result.FromSuccess();
+    }
+
+    public async Task<Result> HandleJoinAsync(Snowflake GuildID, Snowflake UserID, DateTimeOffset joined, bool hasAvatar)
```
Comment (OoLunar) [inline src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs]: Inconsistent parameter naming (GuildId, UserId)

File: src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+            joinRanges.Add(range);
+        }
+
+        if (range.IsSuspect(created))
+        {
+            await _raiders.Writer.WriteAsync(new Raider(GuildID, UserID, "Raid Detection: Join cluster check failed."));
+            return Result.FromSuccess();
+        }
+        
+        //TODO: Check for avatar-less users & calculate risk factor based on account date
```
Comment (OoLunar) [inline src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs:125]: Don't forget about the TODO

File: src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        
+        return Result.FromSuccess();
+    }
+
+    private void TrimBuckets(Snowflake GuildID)
```
Comment (OoLunar) [inline src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs]: Inconsistent parameter naming (GuildId)

**Review by OoLunar — APPROVED** [review]
LGTM, just a bunch of formatting nitpicks

**Review by CalebABG — APPROVED** [review]
Other than a few formatting pieces pointed out by @OoLunar, looks good!

File: src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs (hunk 6 above)
Comment (VelvetToroyashi) [reply src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs:125]: This is more of a "Future me TODO"; it's something that'll be accounted for in V2.

**Pull request merged**